### DISCUSSION TITLE
What is missing for msh3 to be not experimental?

### DISCUSSION THREAD

**Discussion by talregev** [curl/curl#15048] [Q&A] [answered]
Why msh3 is still experimental?

**Comment by icing** [comment, marked as answer]
It is missing vital features that allow us to poll the socket in the right directions to manage progress in an event driven way.

**Comment by talregev** [comment]
@nibanks FYI

**Reply by nibanks** [reply]
IMHO, this design makes a fundamental assumption that (a) there is a socket which is not the case when we use XDP and (b) there is only one which will not be the case for multipath and explicit client migration. For these reasons MsQuic explicitly abstracts this away.

**Comment by talregev** [comment]
Can we mediate with these two approaches?

**Comment by icing** [comment]
libcurl does not need to monitor the socket that actually carries the data. It needs a socket to monitor for events. That could be a notification pipe.

**Reply by nibanks** [reply]
Why not an `epoll_event`? Why does it need to be a socket?

**Reply by icing** [reply]
Ok, to be more precise: a file descriptor that we can "select" on - with "select" being the supported thingie on the platform libcurl is built for.

**Reply by bagder** [reply]
Further: this is partly because the libcurl API *exports* this information to the application so that the event handling can be done outside of libcurl.

**Reply by nibanks** [reply]
Then the curl layer on top of msh3 should create/manage these file descriptors to abstract what it needs to the rest of the library. It's not something that should be exported from msquic or msh3.

**Reply by talregev** [reply]
@icing Can it implement that way? What is missing from msh3 / msquic?

**Reply by bagder** [reply]
It should of course be done there, yes.

**Reply by icing** [reply]
@talregev there are other things which need to be clarified around this. AFAICT there is no information on upload flow control, for example. The application seems to be expected to hold on to the data, but has no way to know when it was actually send and  can release the memory. Nor does it know when new data can be added.

The download parts seems easier to fit into an "event" layer that curl can implement. But it requires threadsafe handling of all those actions, which is a new thing for libcurl to do. So, it's not a small effort.

**Reply by nibanks** [reply]
Every 'send' call has a 'send complete' upcall. On the receive path, for each upcall/event you can either process inline (synchronous completion) or return 'pending' and later call back to 'complete receive'.

**Reply by talregev** [reply]
@nibanks Can msh3 / msquic can transfer the information about the upload / when the data can be added?
Are you also willing to make the effort to help to create such a layer in curl?

Thank you in advance.

**Reply by icing** [reply]
@nibanks where would I find information about that? I seem to have missed this when worked with it last year.

**Reply by nibanks** [reply]
msh3 is effectively a pass-through for msquic. Please see msquic docs for this here: https://github.com/microsoft/msquic/blob/main/docs/Streams.md#sending

**Reply by nibanks** [reply]
Actually, I _think_ curl is still on the older version of msh3 that didn't fully expose these interfaces like msquic. A little bit of work might need to be done to update to the latest first. @talregev I am willing to help here, but likely can't write all the code.

**Reply by icing** [reply]
@nibanks we are on v0.6.0. I saw that you committed some "heavy refactoring". But we cannot follow every commit in your master. Care to tag a new version to look at?

**Reply by talregev** [reply]
@nibanks is it msh3 v0.6.0 is the latest version? If yes, I already added msh3 via vcpkg and curl use it from there.
If v0.6.0 is not the latest, If you can release a newer version tag, and I can added it via vcpkg, then curl can use the latest code.

Can you @icing and @nibanks cooperate about the new layer that needed to add to curl for msh3?

Thank you that you join to this thread.

**Reply by icing** [reply]
@talregev sure, we can cooperate. I cannot promise when and how much time I can commit to that, though.

**Reply by bagder** [reply]
@talregev is there a particular reason for your interest in msh3 support?

**Reply by talregev** [reply]
To make the world one step further toward http3.

**Reply by nibanks** [reply]
Yes, I will tag a new version.

**Reply by nibanks** [reply]
https://github.com/nibanks/msh3/releases/tag/v0.7.0

**Reply by talregev** [reply]
Thank for the work of @dg0yt, vcpkg msh3 port was updated to v0.7.0, and soon will be in curl ci.

@icing @nibanks I hope you can cooperate to create the curl layer on top of msh3 that it needed. 🙏🏻
